evidence/store.go

```diff
@@ -0,0 +1,189 @@
+package evidence
+
+import (
+	"fmt"
+
+	wire "github.com/tendermint/go-wire"
+	"github.com/tendermint/tendermint/types"
+	dbm "github.com/tendermint/tmlibs/db"
+)
+
+/*
+Requirements:
+	- Valid new evidence must be persisted immediately and never forgotten
+	- Uncommitted evidence must be continuously broadcast
+	- Uncommitted evidence has a partial order, the evidence's priority
+
+Impl:
+	- First commit atomically in outqueue, pending, lookup.
+	- Once broadcast, remove from outqueue. No need to sync
+	- Once committed, atomically remove from pending and update lookup.
+		- TODO: If we crash after committed but before removing/updating,
+			we'll be stuck broadcasting evidence we never know we committed.
+			so either share the state db and atomically MarkCommitted
```

Comment:
Doesn't have to be atomic, but if we make updating this evidence store happen first before (or during) state db commit, and make sure the operation on the evidence store is idempotent, then that works too.

Reply:
(would have to comment CONTRACT: needs to be idempotent in that case)